src/main/java/picard/sam/RevertSam.java

```diff
@@ -374,24 +378,46 @@ public void revertSamRecord(final SAMRecord rec) {
                     }
                 }
 
-                // Check that if the first read is marked as unpaired that there is in fact only one read
-                if (!recs.get(0).getReadPairedFlag() && recs.size() > 1) {
-                    log.debug("Discarding " + recs.size() + " reads with name " + recs.get(0).getReadName() + " because they claim to be unpaired.");
-                    discarded += recs.size();
-                    continue readNameLoop;
+                // Get the number of R1s, R2s, and unpaired reads respectively.
+                int firsts = 0, seconds = 0, unpaired = 0;
+                SAMRecord firstRecord = null, secondRecord = null, unpairedRecord = null;
+                for (final SAMRecord rec : recs) {
+                    if (!rec.getReadPairedFlag()) {
+                        if (unpairedRecord == null) unpairedRecord = rec;
+                        ++unpaired;
+                    }
+                    if (rec.getFirstOfPairFlag()) {
+                        if (firstRecord == null) firstRecord = rec;
+                        ++firsts;
+                    }
+                    if (rec.getSecondOfPairFlag()) {
+                        if (secondRecord == null) secondRecord = rec;
+                        ++seconds;
+                    }
                 }
 
-                // Check that if we have paired reads there is exactly one first of pair and one second of pair
-                if (recs.get(0).getReadPairedFlag()) {
-                    int firsts = 0, seconds = 0, unpaired = 0;
-                    for (final SAMRecord rec : recs) {
-                        if (!rec.getReadPairedFlag()) ++unpaired;
-                        if (rec.getFirstOfPairFlag()) ++firsts;
-                        if (rec.getSecondOfPairFlag()) ++seconds;
-                    }
+                // If we have paired reads, then check that there is exactly one first of pair and one second of pair.
+                // Otherwise, check that we have only one unpaired read.
+                if (firsts > 0 || seconds > 0) { // if we have any paired reads
+                    if (firsts != 1 || seconds != 1) { // if we do not have exactly one R1 and one R2
+                        if (KEEP_FIRST_DUPLICATE && firsts >= 1 && seconds >= 1) { // if we have at least one R1 and one R2, we can discard all but the first encountered
+                            discarded += recs.size() - 2;
+                            recs = Arrays.asList(firstRecord, secondRecord);
+                        }  else {
+                            log.debug("Discarding " + recs.size() + " reads with name " + recs.get(0).getReadName() + " because  we found " + firsts + " R1s " + seconds + " R2s and " + unpaired + " unpaired reads.");
```

Comment:
I would just use commas here instead of `+` so that all the string concatenation is only done if the log level is debug.